accepted/2024/unified-vs-settings.md

```diff
@@ -0,0 +1,168 @@
+
+#### Clear Local NuGet Resources
+
+The `Clear local NuGet resources` button from legacy options will be shown as a hyperlink button in USX under the root section's title ("General").
+
+Pressing the hyperlink button will open a modal dialog window indicating the status of the operation of clearing the resources from disk.
+The mockup below shows the button and the effect of pressing it:
+
+![Mockup showing the Clear local NuGet resources hyperlink button and the modal dialog which opens once pressed](../../meta/resources/UnifiedSettings/clear-command-dialog.png)
+
+#### Error messages
+
+When the `NuGet.Config` has an invalid value, the UI will show a blank selection, or an indeterminate state checkbox, and will be accompanied by a simple error message indicating that the value in the backing store was somehow invalid.
+
+Any action taken inside the USX page will overwrite that invalid value in the backing store (`NuGet.Config`) according to `NuGet.Configuration`.
+
+For any fatal errors, such as unable to load the `NuGet.Config` at all due to parsing errors, or similar, the following will occur:
+
+- An error message will be shown under the "General" section label.
+  > "We're unable to read one or more `NuGet.Config` configuration files in this solution. See `(docs link)` for more information."
+  - The `NuGet.Config` will be a hyperlink button which opens the Configuration Files page of NuGet Settings as it is done with the hyperlink, [Settings are stored in NuGet.Config](#settings-are-stored-in-nugetconfig).
+  The shortcut allows the customer to view all paths to configured nuget.config files and open them in the VS editor easily and correct the issue.
+- A retry button is provided by USX.
+- All the edit controls in the section become disabled; the user cannot interact with them while in an error state.
+- The error state can be cleared by if the `NuGet.Config` is modified externally to VS.
+
+#### Help icons
+
+Each UI element in USX can have a help link attached.
+In the mockup, these are shown as a circled `?` icon.
+Pressing one of these opens the relevant Documentation page, and is anchored to the corresponding section in the document.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+### Registration
+
+The NuGet.Client repository will maintain a `registration.json` file to declare all VS Settings, including localizible names, data types, default values, command implementations, and any dependencies among the values of other settings.
+
+The `registration.json` will use the `"type": "external"` mode because our product stores configuration in a custom backing store: One or more `NuGet.Config` files.
+
+### Command: Clear Local Resources
+
+NuGet.Client will create a new modal dialog to indicate that the Clear Local Resources command is in-progress and has completed.
```

Comment:
If it is not possible, I think we should have a message "Clearing local NuGet resources, this might take a minute...". Show progress bar if it is possible.

Reply:
I originally considered a separate notification like you describe; however, it will likely be problematic to allow restores and other NuGet gestures to happen when we are simultaneously deleting from package folders.

Today, the UI freezes while the delete is happening. In this spec, I do not propose freezing the UI in VS. Instead, I propose a modal dialog with a loading indicator. Ideally, we can design a cancellable Locals Command so that a cancel button can be shown on this dialog.

Reply:
ok, let's proceed with this as MVP and see if we have any feedback from users